### DISCUSSION TITLE
Split package add/remove  management and  personal config management

### DISCUSSION THREAD

**Discussion by lld2001** [emacscollective/borg#118] [Ideas]
Like [doom-emacs](https://github.com/hlissner/doom-emacs), which seperate package add/remove  management from whole .emacs.d config management. I think that's because add/remove package is not so frequently. Doom-emacs will generate a big `autoloads.el` file when changing packages.  

Could `borg` make the same improvement ? Because it can speed up startup time especially in Windows(forking `git` process is very slow)?

**Comment by tarsius** [comment]
Years before I published `borg` it used to do that for a while, but I removed it because I felt it only complicated matters. Once could no longer just remove a submodule manually but had to always use `borg-remove` because otherwise the obsolete autoload definitions would not get cleaned out. I also checked what the competition did; `el-get` used a single autoloads file, but `package` also used one per package, making it easier to go with that approach. I believe this is still the case.

I don't plan to implement this any time soon if ever. But you could implement a command that removes the per-package autoload files and instead generates a single file. You of course have to remember that command when appropriate (though forgetting to do so would not be a big deal, then those per-package autoload files would just also get loaded). And you would have to load the big autoload file explicitly in your init file.

> (forking `git` process is very slow)

Luckily loading the autoload files doesn't involve calling git, so the only slow down is due to having to open N files instead of one.

**Comment by lld2001** [comment]
> Once could no longer just remove a submodule manually but had to always use `borg-remove`

I found both methods lead to modifying the `.gitmodules` file. So it may use this fact when run `borg-initialize`. I think that it can compare the timestamp of the `.gitmodules` and the `big autoloads` file, and then decide whether to regenerate the `big autoloads` file or simply load it.

> Luckily loading the autoload files doesn't involve calling git, so the only slow down is due to having to open N files instead of one.

Sorry for this. I actually mean `borg-initialize` calling `git` process many times. If the `big autoloads` file can achieve the same task as the `borg-initialize` function. Then the normal process of loading packages may not depend on `git` process. Some task of the `borg-initialize` function may be replaced by `big autoloads` file.

I eagerly need this because there's an extra problem. I have an `offline` computer which need frequently to sync configuration with the other `online` computer(can connect to Internet). Right now, I have to maintain a backup of `.emacs.d` which is the same as the `offline` computer. When I add/remove packages in `online` computer, I have to compare the changes in `.git` directory. This process involves many trivial details.

**Reply by lld2001** [reply]
I've made this `single autoloads` method work by the following snippets. Using these codes to replace simply `borg-initialize`, it will generate a big autoload file in the first time. Then the following startup of Emacs can load this file instead of `borg-initialize`. Which will gain a minor speedup and reduce the git process call in `borg-initialize`. 

```emacs-lisp
(eval-and-compile ; `borg'
  (add-to-list 'load-path (expand-file-name "lib/borg" user-emacs-directory))
  (require 'borg)

  (defun lld-collect-autoloads (file)
    "insert all enabled drone's autoloads file to a single file."
    (make-directory (file-name-directory file) 'parents)
    (condition-case-unless-debug e
        (with-temp-file file
          (setq-local coding-system-for-write 'utf-8)
          (let ((standard-output (current-buffer))
                (print-quoted t)
                (print-level nil)
                (print-length nil)
                path lpath auto)
            (insert ";; -*- lexical-binding: t; coding: utf-8; no-native-compile: t -*-\n"
                    ";; This file is generated from enabled drones.\n")
            (prin1 `(set `load-path ',load-path))
            (insert "\n")
            (print `(set `custom-theme-load-path ',custom-theme-load-path))
            (insert "\n")

            ;; insert all drone's autoloads.el to this file
            (dolist (drone (nreverse (borg-drones 'raw)))
              (setq path (car (plist-get (cdr drone) 'path)))
              (setq drone-path (concat borg-top-level-directory path))
              (when (file-exists-p drone-path)
                (setq auto (car(directory-files drone-path t ".*-autoloads.el\\'")))
                (when (and auto
                           (file-exists-p auto))
                  (insert-file-contents auto))))t))
      (error (delete-file file)
             (signal 'collect-autoload-error (list file e)))))

  (defvar single-autoload-path "~/.emacs.d/etc/borg/autoload/" "single autoload file.")
  (let ((file (concat single-autoload-path
                      "autoload-"
                      (format-time-string
                       "%+4Y%m%d-%H%M%S"
                       (file-attribute-modification-time
                        (file-attributes borg-gitmodules-file)))
                      ".el")))
    (if (file-exists-p file)
        (load file)
      (borg-initialize)
      (lld-collect-autoloads file))))
```
